src/nxt_array.c

```diff
@@ -1,150 +1,253 @@
 
 /*
+ * Copyright (C) Evgenii Sokolov
  * Copyright (C) Igor Sysoev
  * Copyright (C) NGINX, Inc.
  */
 
 #include <nxt_main.h>
 
 
+static void nxt_array_elts_copy(void **dst_elts, nxt_array_t *src_array);
+static nxt_uint_t nxt_array_calc_alloc(nxt_uint_t nelts);
+static void *nxt_array_elts_realloc(nxt_array_t *array,
+                                    nxt_uint_t nalloc);
+
+
+/*
+ * nxt_array_create - Create a new array and return a pointer to it.
+ */
 nxt_array_t *
-nxt_array_create(nxt_mp_t *mp, nxt_uint_t n, size_t size)
+nxt_array_create(nxt_mp_t *mp, nxt_uint_t arr_size, size_t elt_size)
 {
-    nxt_array_t  *array;
+    nxt_uint_t     nalloc;
+    nxt_array_t    *array;
+    void           *elts;
 
-    array = nxt_mp_alloc(mp, sizeof(nxt_array_t) + n * size);
+    array = nxt_mp_alloc(mp, sizeof(nxt_array_t));
 
     if (nxt_slow_path(array == NULL)) {
         return NULL;
     }
 
-    array->elts = nxt_pointer_to(array, sizeof(nxt_array_t));
+    nalloc = nxt_array_calc_alloc(arr_size);
+    elts = nxt_mp_alloc(mp, elt_size * nalloc);
+
+    if (nxt_slow_path(elts == NULL)) {
+        return NULL;
+    }
+
+    array->elts = elts;
     array->nelts = 0;
-    array->size = size;
-    array->nalloc = n;
+    array->size = elt_size;
+    array->nalloc = nalloc;
     array->mem_pool = mp;
 
     return array;
 }
 
 
+/*
+ * nxt_array_destroy - Destroy the existing array.
+ */
 void
 nxt_array_destroy(nxt_array_t *array)
 {
-    if (array->elts != nxt_pointer_to(array, sizeof(nxt_array_t))) {
-        nxt_mp_free(array->mem_pool, array->elts);
-    }
+    if (nxt_fast_path(array != NULL)) {
+
+        array->nelts = 0;
 
-    nxt_mp_free(array->mem_pool, array);
+        (void) nxt_mp_free(array->mem_pool, array->elts);
+        (void) nxt_mp_free(array->mem_pool, array);
+
+    }
 }
 
 
+/*
+ * nxt_array_add - Add an element to the existing array.
+ */
 void *
 nxt_array_add(nxt_array_t *array)
 {
-    void      *p;
-    uint32_t  nalloc, new_alloc;
-
-    nalloc = array->nalloc;
+    if (nxt_fast_path(array != NULL)) {
 
-    if (array->nelts == nalloc) {
+        nxt_uint_t    nalloc;
+        void          *elt, *elts;
 
-        if (nalloc < 16) {
-            /* Allocate new array twice larger than current. */
-            new_alloc = (nalloc == 0) ? 4 : nalloc * 2;
```

Comment:
- fixed the formula for calculating the required amount of memory (nalloc) allocated for the created array;

4 is too big for an empty and possibly never used array in the future.

Reply:
Ignoring if this change is good or not, which should need much more investigation from my side...

Please, when submitting changes that are logically different from each other, submit them as small separate commits.  Even if it all makes sense together as a cleanup (so everything makes sense as a pull request), each small change makes sense separately as a single commit.  It is much easier to review them that way, and check correctness of each small bit.  That allows that some changes may be rejected while others may be cherry-picked.